src/DataTypes/convertYTsaurusDataType.cpp

```diff
@@ -0,0 +1,366 @@
+    if (precision <=  DataTypeDecimalBase<Decimal32>::maxPrecision())
+        data_type_ptr = std::make_shared<DataTypeDecimal<Decimal32>>(precision, scale);
+    else if (precision <= DataTypeDecimalBase<Decimal64>::maxPrecision())
+        data_type_ptr = std::make_shared<DataTypeDecimal<Decimal64>>(precision, scale);
+    else if (precision <= DataTypeDecimalBase<Decimal128>::maxPrecision())
+        data_type_ptr = std::make_shared<DataTypeDecimal<Decimal128>>(precision, scale);
+    else if (precision <= DataTypeDecimalBase<Decimal256>::maxPrecision())
+        data_type_ptr = std::make_shared<DataTypeDecimal<Decimal256>>(precision, scale);
+    return data_type_ptr;
+}
+
+DataTypePtr convertYTOptional(const Poco::JSON::Object::Ptr & json)
+{
+    if (!json->has("item"))
+    {
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Couldn't parse 'optional' type from YT('item' was not found)");
+    }
+    auto nested_type = convertYTItemType(json->get("item"));
+    if (!nested_type)
+    {
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Couldn't parse 'optional' type from YT(incorrect nested type)");
+    }
+    if (nested_type->isNullable() || !nested_type->canBeInsideNullable())
+    {
+        return nested_type;
+    }
+    return std::make_shared<DataTypeNullable>(nested_type);
+}
+
+DataTypePtr convertYTList(const Poco::JSON::Object::Ptr & json)
+{
+    return std::make_shared<DataTypeArray>(convertYTItemType(json->get("item")));
+}
+
+DataTypePtr convertYTTuple(const Poco::JSON::Object::Ptr & json)
+{
+    DataTypes types;
+    auto tuple_types = json->getArray("elements");
+    types.reserve(tuple_types->size());
+    for (const auto & tuple_type : *tuple_types)
+    {
+        auto element_json = tuple_type.extract<Poco::JSON::Object::Ptr>();
+        types.push_back(convertYTItemType(element_json->get("type")));
+    }
+    return std::make_shared<DataTypeTuple>(types);
+}
+
+DataTypePtr convertYTDict(const Poco::JSON::Object::Ptr &json)
+{
+    auto key = convertYTItemType(json->get("key"));
+    auto value = convertYTItemType(json->get("value"));
+    return std::make_shared<DataTypeMap>(key, value);
+}
+
+DataTypePtr convertYTVariant(const Poco::JSON::Object::Ptr &json)
+{
+    if (!json->has("elements"))
+    {
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Couldn't parse 'variant' type from YTsaurus");
+    }
+    auto elements_json = json->getArray("elements");
+    DataTypes types;
+    types.reserve(elements_json->size());
+    for (const auto & element : *elements_json)
+    {
+        auto element_json = element.extract<Poco::JSON::Object::Ptr>();
+        types.push_back(convertYTItemType(element_json->get("type")));
+    }
+    return std::make_shared<DataTypeVariant>(types);
+}
+
+DataTypePtr convertYTStruct(const Poco::JSON::Object::Ptr &json)
+{
+    if (!json->has("members"))
+    {
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Couldn't parse 'struct' type from YTsaurus");
```

Comment:
```suggestion
        throw Exception(ErrorCodes::INCORRECT_DATA, "Couldn't parse 'struct' type from YTsaurus. Missing field `member`.");
```
For exception below pls add more details in the msg.